### PR TITLE
Add `Verify` trait to `mc-sgx-dcap-quoteverify`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
Adds a Verify to be used with DCAP Quote types

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #128** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/128" target="_blank">Graphite</a> 
    * **PR #137** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/137" target="_blank">Graphite</a> 
      * **PR #138** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/138" target="_blank">Graphite</a> 
        * **PR #139** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/139" target="_blank">Graphite</a> 
          * **PR #140** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/140" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/140?utm_source=stack-comment).

File: core/types/src/quote.rs [hunk 1]
```diff
@@ -112,6 +112,12 @@ pub struct RawQuote<'a> {
     bytes: &'a [u8],
 }
 
+impl<'a> AsRef<[u8]> for RawQuote<'a> {
```
Comment (nick-mobilecoin) [inline core/types/src/quote.rs:115]: not sure I like the AsRef.  I considered implementing a ptr and length method, but it didn't really seem to provide more over `AsRef`.

File: dcap/quoteverify/src/verify.rs [hunk 2]
```diff
@@ -2,8 +2,43 @@
...
+use mc_sgx_dcap_sys_types::{sgx_ql_qv_result_t, time_t};
 use mc_sgx_dcap_types::Quote3Error;
 use mc_sgx_util::ResultInto;
+use std::{
+    ptr,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+pub trait Verify: BaseQuote {
+    fn verify(&self) -> Result<(), Quote3Error> {
```
Comment (nick-mobilecoin) [inline dcap/quoteverify/src/verify.rs:15]: This is _minimal_ exercising of the verification functionality.  I'm up for suggestions on how to further extend.  It feels like a builder pattern might be nice `with_supplemental().with_collateral().with_report_info().verify()`

File: dcap/types/src/quote3.rs [hunk 3]
```diff
@@ -366,6 +366,87 @@ impl From<Error> for quote3_error_t {
...
 
+/// An enumeration of errors which occur when verifying a quote
+///
+/// These errors correspond to error elements of
+/// [`sgx_ql_qv_result_t`](mc_sgx_dcap_sys_types::sgx_ql_qv_result_t).
+#[derive(Copy, Clone, Debug, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[non_exhaustive]
+#[repr(u32)]
+pub enum VerificationError {
```
Comment (nick-mobilecoin) [inline dcap/types/src/quote3.rs:377]: Not sure if we want this in a separate module, it is for to quote version 3...

File: dcap/quoteverify/src/verify.rs [hunk 4]
```diff
@@ -2,8 +2,43 @@
...
+        let mut quote_status = sgx_ql_qv_result_t::SGX_QL_QV_RESULT_MAX;
+
+        unsafe {
+            mc_sgx_dcap_quoteverify_sys::sgx_qv_verify_quote(
+                raw_quote.as_ref().as_ptr(),
+                raw_quote.as_ref().len() as u32,
+                ptr::null(),
+                since_the_epoch.as_secs() as time_t,
+                &mut expiration_status,
+                &mut quote_status,
```
Comment (nick-mobilecoin) [inline dcap/quoteverify/src/verify.rs:31]: up for suggestions on how to best chain these results and how best to prioritize which return value should win, or if we should pass some kind of error struct

File: core/types/src/quote.rs (hunk 1 above)
Comment (awygle) [reply core/types/src/quote.rs:115]: What's the use case for this impl? I am wondering if maybe what we want is a cheap `Serialize` implementation...

File: dcap/types/src/quote3.rs (hunk 3 above)
Comment (awygle) [reply dcap/types/src/quote3.rs:377]: I think it's fine here personally.

File: dcap/quoteverify/src/verify.rs (hunk 2 above)
Comment (awygle) [reply dcap/quoteverify/src/verify.rs:15]: I agree that a builder pattern is the way to go here.

File: dcap/quoteverify/src/verify.rs (hunk 4 above)
Comment (awygle) [reply dcap/quoteverify/src/verify.rs:31]: I don't really understand what you mean by "which return value should win"?

**Review by awygle — APPROVED** [review]
I think this is fine for a start, I do think we should do a builder pattern later on to include advanced functionality.

**Review by jcape — APPROVED** [review]

**Comment by jcape** [comment]
Generally agree, but I do have to ask, what's the use case for this stuff? Like, where do we expect to use it?

**Comment by nick-mobilecoin** [comment]
Closing this out.
The desire is to utilize common verify functionality that can be used consistently inside and outside of the enclave.  This means that the logic will not directly rely on SGX

**Pull request closed without merging**